pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,698 @@
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageosMounter
+	pvUID types.UID
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageosMounter
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	adminSecretName := ""
+	adminSecretNamespace := "default"
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "adminsecretname":
+			adminSecretName = v
+		case "adminsecretnamespace":
+			adminSecretNamespace = v
+		case "volumenamespace":
+			c.volNamespace = v
+		case "description":
+			c.description = v
+		case "pool":
+			c.pool = v
+		case "fstype":
+			c.fsType = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+		}
+	}
+
+	// Set from PVC
+	c.podNamespace = c.options.PVC.Namespace
+	c.volName = c.options.PVName
+	if c.volNamespace == "" {
+		c.volNamespace = c.options.PVC.Namespace
+	}
+	c.labels = make(map[string]string)
+	for k, v := range c.options.PVC.Labels {
+		c.labels[k] = v
+	}
+	capacity := c.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	c.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	if adminSecretName == "" {
+		return nil, fmt.Errorf("missing StorageOS admin secret name")
+	}
+
+	apiCfg, err := parsePVSecret(adminSecretNamespace, adminSecretName, c.plugin.host.GetKubeClient())
+	if err != nil {
+		return nil, err
+	}
+	c.apiCfg = apiCfg
+
+	vol, err := c.manager.CreateVolume(c)
+	if err != nil {
+		glog.Errorf("failed to create volume: %v", err)
+		return nil, err
+	}
+	if vol.FSType == "" {
+		vol.FSType = defaultFSType
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   vol.Name,
+			Labels: map[string]string{},
+			Annotations: map[string]string{
+				volumehelper.VolumeDynamicallyCreatedByKey: "storageos-dynamic-provisioner",
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: c.options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   c.options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", vol.Size)),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				StorageOS: &v1.StorageOSVolumeSource{
+					VolumeName:      vol.Name,
+					VolumeNamespace: vol.Namespace,
+					FSType:          vol.FSType,
+					ReadOnly:        false,
+					SecretRef: &v1.LocalObjectReference{
+						Name: adminSecretName,
```

Comment:
> To confirm, I should change the LocalObjectReference to an ObjectReference with Namespace set to adminSecretNamespace, then update the Visitor functions to add SecretRef.Namespace?

That would give the node running the mount/unmount operation for the PV the info it needs, and the node authorizer the info it needs, so it makes sense to me.
* cc @kubernetes/sig-storage-api-reviews on the data flow for the storage plugin
* cc @kubernetes/sig-auth-api-reviews on the data flow for the node authorizer

I would expect a storage class to specify both secret name and namespace, or neither (to use default storageos credentials). If just the secret name or namespace, I'd expect the dynamic provision to fail, not fall back to looking in the `default` namespace.

Also, this is the first volume source I know of that differs between the PV spec and the Pod spec (pod spec would only allow a local secret reference, pv spec would allow a fully specified name/namespace reference). That makes sense to me, given the use, but I wanted to call it out.